### PR TITLE
Alex

### PR DESCRIPTION
**Pull request by AlexCozma**
Done

### CODE REVIEW

File: MySolution.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
+class MySolution
+
+    #problem1
+    def sumOfEven(array)
+        sum = 0
```
Comment (alexandru-calinoiu) [inline MySolution.rb]: So c#/java like :) can you come up with a 1 liner?

Steps:
- how can you drop `sum = 0` and `return sum`? do you even need a `return` statement in ruby?
- what are some alternatives to `for`?
- condense code.

File: MySolution.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+            if i.even?
+                sum += i
+            end
+        end
+        return sum
+    end
+    
+    #problem2
+    def getLargest2Numbers(array)
+        bigMax = -1
```
Comment (alexandru-calinoiu) [inline MySolution.rb]: The problem states `an array of integers` does not say `unsigned integers`. So this solution will fail for an array like `[-3, -2]`

File: MySolution.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
+class MySolution
+
+    #problem1
+    def sumOfEven(array)
+        sum = 0
+        for i in array
+            if i.even?
```
Comment (alexandru-calinoiu) [inline MySolution.rb]: :+1: for using `even?`

File: MySolution.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        end
+        return sum
+    end
+    
+    #problem2
+    def getLargest2Numbers(array)
+        bigMax = -1
+        smallMax = -1
+        for i in array
+            if i > bigMax and i > smallMax #if bigger then both
```
Comment (alexandru-calinoiu) [inline MySolution.rb]: This comments don't actually bring any benefit. Please write the code in such a way that it reads clearly and it does not require any explanation.

File: MySolution.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        end
+        result = [smallMax, bigMax]
+        return result
+    end
+
+    #problem3
+    def hasDublicates(array)
+        #without temporary data structure to help, or sorting the problem
+        #goes to compare each element with each other
+        for i in 0..array.length * array.length
```
Comment (alexandru-calinoiu) [inline MySolution.rb]: Very cool solution.

But you technically cheated `Try to find a solution by executing just one loop.` you masked a loop in a loop with your 2 indexes.

**Review by alexandru-calinoiu — CHANGES_REQUESTED** [review]
2 solutions are still pending and 1 can be improved.

**Comment by levelboy** [comment]
@AlexCozma Before I even looked at the solution, I took a look over the tests. That doesn't look like TDD. 
There are three rules of TDD that drive the red green-refactor-cycle (taken from here: http://butunclebob.com/ArticleS.UncleBob.TheThreeRulesOfTdd)

1. You are not allowed to write any production code unless it is to make a failing unit test pass.
2. You are not allowed to write any more of a unit test than is sufficient to fail; and compilation failures are failures.
3. You are not allowed to write any more production code than is sufficient to pass the one failing unit test.

Please have a look at this video https://www.youtube.com/watch?v=qkblc5WRn-U 
See how the tests evolve from very simple to more complex, along as the code evolves also.

File: MySolution.rb (hunk 1 above)
Comment (AlexCozma) [reply MySolution.rb]: Well I admit that C# is may favorite language :) and Ruby is truly new to me and I just got started learning it - And Yeah I can try to make it in one line.

- For dropping `sum = 0` and the `return sum` I can use an array function from ruby `array.sum` and to sum over only the even values something like `array.select{|v| v.even?}` (that looks like a LINQ from C#)
- If you referring to loop alternatives to `for` in ruby I know about `while`, `until`,`begin-while`

File: MySolution.rb (hunk 3 above)
Comment (AlexCozma) [reply MySolution.rb]: Thanks :)

File: MySolution.rb (hunk 2 above)
Comment (AlexCozma) [reply MySolution.rb]: You are right, I must had in mind the hypothesis from problem 3 after I read them. I'll add this Test Case and correct the solution

File: MySolution.rb (hunk 4 above)
Comment (AlexCozma) [reply MySolution.rb]: Ok, I understood

File: MySolution.rb (hunk 1 above)
Comment (AlexCozma) [reply MySolution.rb]: Well I admit that C# is may favorite language :) and Ruby is truly new to me and I just got started learning it - And Yeah I can try to make it in one line.

-For dropping `sum = 0` and the return sum I can use an array function from ruby `array.sum` and to sum over only the even values something like `array.select{|v| v.even?}` (that looks like a LINQ from C#)
-If you referring to loop alternatives to for in ruby I know about `while`, `until`,`begin-while`

File: MySolution.rb (hunk 2 above)
Comment (AlexCozma) [reply MySolution.rb]: You're right, I've must had in mind the hypothesis from problem 3 after I read the problems. I'll correct it

File: MySolution.rb (hunk 3 above)
Comment (AlexCozma) [reply MySolution.rb]: Thanks :)

File: MySolution.rb (hunk 4 above)
Comment (AlexCozma) [reply MySolution.rb]: Actually yeah, the comment was redundant

File: MySolution.rb (hunk 5 above)
Comment (AlexCozma) [reply MySolution.rb]: Thanks. 
Yes I'm aware that is technically a cheat, as I wrote in the comment I couldn't see anyway how to compare each element with each other. I'll try to find a different approach

File: .vscode/launch.json [hunk 6]
```diff
@@ -0,0 +1,71 @@
+{
```
Comment (alexandru-calinoiu) [inline .vscode/launch.json:1]: I don't believe this file should have been commited.

File: MySolution.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    def sumOfEven(array)
+        array.select{|v| v.even?}.sum
+    end
+    
+    #problem2
+    def getLargest2Numbers(array)
+        bigMax = -1/0.0
+        smallMax = -1/0.0
+        for i in array
+            if i > bigMax and i > smallMax
```
Comment (alexandru-calinoiu) [inline MySolution.rb]: Using `and` comes with a warning https://stackoverflow.com/questions/1426826/difference-between-and-and-in-ruby

File: MySolution.rb [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    
+    #problem2
+    def getLargest2Numbers(array)
+        bigMax = -1/0.0
+        smallMax = -1/0.0
+        for i in array
+            if i > bigMax and i > smallMax 
+                smallMax = bigMax
+                bigMax = i  
+            else
```
Comment (alexandru-calinoiu) [inline MySolution.rb]: Syntax can be shorten with `elsif`

File: MySolution.rb [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+                if i > smallMax 
+                    smallMax = i 
+                end
+            end
+        end
+        result = [smallMax, bigMax]
+        return result
```
Comment (alexandru-calinoiu) [inline MySolution.rb]: Still using `return`'s :(

File: MySolution.rb [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    end
+
+    #problem3
+    def hasDublicates(array)
+        for i in 0..array.length 
+            if array.select {|v| v == array[i]}.size > 1
```
Comment (alexandru-calinoiu) [inline MySolution.rb]: Still not right, `select` uses a loop in it's implementation therefor you still have a inner loop.

File: MySolution.rb [hunk 11]
```diff
@@ -0,0 +1,36 @@
+class MySolution
+
+    #problem1
+    def sumOfEven(array)
+        array.select{|v| v.even?}.sum
```
Comment (alexandru-calinoiu) [inline MySolution.rb:5]: This looks like ruby now :+1: 

You can make it even nicer take a look at `inject`

File: MySolutionUnitTest.rb [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    end
+
+#problem 3
+    def test_hasDublicates_true
+        array = [5, 9, 1, 18, 6, 25, 16, 6]
+        assert_equal true,@unit.hasDublicates(array)
+    end
+
+    def test_hasDublicates_false
+        array = [5, 9, 1, 18, 6, 25, 16, 8]
```
Comment (alexandru-calinoiu) [inline MySolutionUnitTest.rb]: Your test case is not valid, the problem states `Given an array of numbers having size N in which every number is between 1 and N`

File: MySolution.rb (hunk 7 above)
Comment (levelboy) [inline MySolution.rb]: Do you need the second condition here?

File: MySolution.rb [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    end
+    
+    #problem2
+    def getLargest2Numbers(array)
+        bigMax = -1/0.0
+        smallMax = -1/0.0
+        for i in array
```
Comment (levelboy) [inline MySolution.rb]: Ruby code should be more expressive. 
Confusion here is that usually, when you do a `for`, the `i` is the index. In this case, it's the actual item in the array.

File: MySolution.rb (hunk 9 above)
Comment (levelboy) [inline MySolution.rb]: Do you need a variable named result here?
Do you need to explicitly specify `return` in ruby? (this was a previous question from @alexandru-calinoiu )

File: MySolution.rb (hunk 10 above)
Comment (levelboy) [inline MySolution.rb]: This is still O(n^2)

File: MySolution.rb [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+                end
+            end
+        end
+        result = [smallMax, bigMax]
+        return result
+    end
+
+    #problem3
+    def hasDublicates(array)
+        for i in 0..array.length
```
Comment (levelboy) [inline MySolution.rb]: Emphasizing the confusion again for how you use `i`. This time, it's an item in an array of indices, used to access the array in the inner loop.

File: MySolution.rb [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+        end
+        result = [smallMax, bigMax]
+        return result
+    end
+
+    #problem3
+    def hasDublicates(array)
+        for i in 0..array.length 
+            if array.select {|v| v == array[i]}.size > 1
+                return true
```
Comment (levelboy) [inline MySolution.rb]: See how `return` works in a ruby function

File: MySolutionUnitTest.rb [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+    end
+
+    def test_getLargest2Numbers_negativeValues
+        array = [-3, -2]
+        assert_equal array,@unit.getLargest2Numbers(array)
+    end
+
+#problem 3
+    def test_hasDublicates_true
+        array = [5, 9, 1, 18, 6, 25, 16, 6]
```
Comment (levelboy) [inline MySolutionUnitTest.rb]: The input to the 3rd problem does not respect the problem statement:
`Given an array of numbers having size N in which every number is between 1 and N`

File: .vscode/launch.json (hunk 6 above)
Comment (levelboy) [inline .vscode/launch.json:1]: What's up with this file? :)

**Review by levelboy — CHANGES_REQUESTED** [review]
Please see requested changes.

**Review by alexandru-calinoiu — CHANGES_REQUESTED** [review]

File: MySolution.rb (hunk 7 above)
Comment (AlexCozma) [reply MySolution.rb]: Giving a second thought, yes it's actually redundant.

File: MySolution.rb (hunk 13 above)
Comment (AlexCozma) [reply MySolution.rb]: Yes, it is, this if from my habit that first choice for a variable in for to be named `i`. In this case it is confusing, I'll give it a suggestive name

File: MySolution.rb (hunk 9 above)
Comment (AlexCozma) [reply MySolution.rb]: In ruby, no I don't need a explicit result. I didn't know how to return two variables at once, that's why I used a variable. But yes I can leave it as `[smallMax, bigMax]`

File: MySolutionUnitTest.rb (hunk 16 above)
Comment (AlexCozma) [reply MySolutionUnitTest.rb]: You're right I skip notice of it, I'll correct it

File: .vscode/launch.json (hunk 6 above)
Comment (AlexCozma) [reply .vscode/launch.json:1]: It's the debugger configuration file from visual studio code, is generated in the workspace folder where also is my local repository. Sorry for that

Comment (AlexCozma) [reply .vscode/launch.json:1]: It's the debugger configuration file from visual studio code, is generated in the workspace folder where also is my local repository. Sorry for that

File: MySolution.rb (hunk 7 above)
Comment (AlexCozma) [reply MySolution.rb]: I didn't know that. Still the second condition is redundant

File: MySolution.rb (hunk 8 above)
Comment (AlexCozma) [reply MySolution.rb]: OK I'll change it

File: MySolution.rb (hunk 9 above)
Comment (AlexCozma) [reply MySolution.rb]: I'll remove it

File: MySolutionUnitTest.rb (hunk 12 above)
Comment (AlexCozma) [reply MySolutionUnitTest.rb]: I'll correct it.

File: MySolutionUnitTest.rb [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+    end
+
+    def test_getLargest2Numbers_BigMaxFirst
+        array = [5, 9, 1, 19, 14, 3, 10, 25, 9]
+        result = [19,25]
+        assert_equal result,@unit.getLargest2Numbers(array)
+    end
+
+    def test_getLargest2Numbers_negativeValues
+        array = [-3, -2]
```
Comment (levelboy) [inline MySolutionUnitTest.rb:41]: Does this test work if you put set the array to: [-2, -3]?

Comment (AlexCozma) [reply MySolutionUnitTest.rb:41]: Yes it passes

**Pull request closed without merging**